jscomp/syntax/ast_utf8_string.ml

```diff
@@ -142,4 +141,102 @@ and unicode loc buf s offset s_len =
 (* http://www.2ality.com/2015/01/es6-strings.html
    console.log('\uD83D\uDE80'); (* ES6*)
    console.log('\u{1F680}');
-*)   
\ No newline at end of file
+*)
+
+type interpo = Text of string | Delim of string
+
+let consume_text s start_index = 
+  let rec _consume_text s index last_char new_word =
+    if index = String.length s then new_word, String.length s
+    else begin 
+     match s.[index] with
+     | '$' -> if last_char = '\\' then _consume_text s (index+1) '$' (Ext_string.append new_word '$') 
+              else (new_word, index)
+     | c -> _consume_text s (index + 1) c (Ext_string.append new_word c)
+     end 
+  in _consume_text s start_index ' ' "" 
+ 
+let consume_delim s start_index =
+  let with_par = ref false in
+  let rec _consume_delim s index ident =
+    if index = String.length s then (if !with_par = true then (None, index) else (Some ident, index))
+    else
+      match s.[index] with
+      | '(' -> (if !with_par = false then (with_par := true; _consume_delim s (index+1) ident) else (None, index))
+      | ')' -> (if !with_par = false then (None, index + 1) else (with_par := false; (Some ident, index+1)))
+      | '$' -> (_consume_delim s (index+1) ident)
+      | c -> if (Char.code c >= Char.code 'a' && Char.code c <= Char.code 'z') || 
+                (Char.code c >= Char.code 'A' && Char.code c <= Char.code 'Z') ||
+                (Char.code c >= Char.code '0' && Char.code c <= Char.code '9') ||
+                Char.code c = Char.code '_'
+             then _consume_delim s (index+1) (Ext_string.append ident c)
+             else if !with_par = false then (Some ident, index) else (None, index + 1) 
+  in match s with 
+  | "" -> (Some "", start_index)
+  | _ -> if start_index = String.length s then (Some "", start_index)
+         else (if s.[start_index] <> '$' then (None, start_index)
+           else _consume_delim s start_index "")
+
+
+let compute_new_loc (loc:Location.t) s = let length = String.length s in 
+  let new_loc = 
+    {loc with loc_start = {loc.loc_start with pos_cnum = loc.loc_end.pos_cnum}; 
+              loc_end = {loc.loc_start with pos_cnum = loc.loc_end.pos_cnum + length}} 
+  in new_loc
+
+let error_reporting_loc (loc:Location.t) start_index end_index =
+  let new_loc =
+    {loc with loc_start = {loc.loc_start with pos_cnum = loc.loc_start.pos_cnum + start_index};
+              loc_end   = {loc.loc_end   with pos_cnum = loc.loc_start.pos_cnum + end_index }} in new_loc
+
+let split_es6_string s loc = 
+  let rec _split s index nl = 
+    if index >= String.length s then List.rev nl 
+    else begin
+      match consume_text s index, consume_delim s index with
+      | ("" , str_index)  , (None   , err_index) -> let new_loc = error_reporting_loc loc index err_index in Location.raise_errorf ~loc:new_loc "Not a valid es6 template string"
```

Comment:
instead of raising directly, can we return an optional for easy unit-testing?
The current unit testing depends on compiler-libs, it would be nice that such utility is self-contained